src/ol/render/canvas/Immediate.js

```diff
@@ -614,7 +622,17 @@ class CanvasImmediateRenderer extends VectorContext {
    */
   drawFeature(feature, style) {
     const geometry = style.getGeometryFunction()(feature);
-    if (!geometry || !intersects(this.extent_, geometry.getExtent())) {
+    if (!geometry) {
+      return;
+    }
+    let drawGeometry = geometry;
+    if (this.squaredTolerance_) {
+      drawGeometry = geometry.simplifyTransformed(
+        this.squaredTolerance_,
+        this.userTransform_
+      );
+    }
+    if (!intersects(this.extent_, drawGeometry.getExtent())) {
```

Comment:
This is already done by the `drawGeometry()` function which is called below, so no need to make any change here.